src/Storages/Hive/HiveFilesCollector.cpp

```diff
@@ -0,0 +1,270 @@
+#include <Storages/Hive/HiveFilesCollector.h>
+#if USE_HIVE
+#include <Formats/FormatFactory.h>
+#include <Interpreters/ExpressionAnalyzer.h>
+#include <Interpreters/TreeRewriter.h>
+#include <Parsers/ASTFunction.h>
+#include <Processors/Executors/PullingPipelineExecutor.h>
+#include <Common/ErrorCodes.h>
+#include <Processors/Formats/IInputFormat.h>
+#include <QueryPipeline/Pipe.h>
+#include <QueryPipeline/QueryPipeline.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <base/logger_useful.h>
+#include <IO/ReadBufferFromString.h>
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int INVALID_PARTITION_VALUE;
+}
+ASTPtr HiveFilesCollector::extractKeyExpressionList(const ASTPtr & node)
+{
+    if (!node)
+        return std::make_shared<ASTExpressionList>();
+    const auto * expr_func = node->as<ASTFunction>();
+    if (expr_func && expr_func->name == "tuple")
+        return expr_func->arguments->clone();
+
+    auto res = std::make_shared<ASTExpressionList>();
+    res->children.push_back(node);
+    return res;
+}
+void HiveFilesCollector::prepare()
+{
+    ASTPtr partition_key_expr_list = extractKeyExpressionList(partition_by_ast);
+    NamesAndTypesList all_name_and_types = columns.getAllPhysical();
+    if (!partition_key_expr_list->children.empty())
+    {
+        auto syntax_result = TreeRewriter(context).analyze(partition_key_expr_list, all_name_and_types);
+        partition_key_expr = ExpressionAnalyzer(partition_key_expr_list, syntax_result, context).getActions(false);
+        partition_name_and_types = partition_key_expr->getRequiredColumnsWithTypes();
+        partition_minmax_idx_expr = std::make_shared<ExpressionActions>(
+            std::make_shared<ActionsDAG>(partition_name_and_types), ExpressionActionsSettings::fromContext(context));
+    }
+
+    for (const auto & column : all_name_and_types)
+    {
+        if (partition_name_and_types.contains(column.name))
+            hive_file_name_and_types.push_back(column);
+    }
+
+    hive_file_minmax_idx_expr = std::make_shared<ExpressionActions>(
+        std::make_shared<ActionsDAG>(hive_file_name_and_types), ExpressionActionsSettings::fromContext(context));
+}
+std::vector<HiveFilesCollector::FileInfo> HiveFilesCollector::collect(PruneLevel prune_level)
```

Comment:
Why do we  need to return `std::vector<HiveFilesCollector::FileInfo>` instead of `std::vector<HiveFilePtr>`.

Reply:
`HiveFilesCollector::FileInfo` contains more information that will be needed in `StorageHiveCluster`

Since https://github.com/ClickHouse/ClickHouse/pull/35164 has no process for a quit long time, but we still have some optimization on `StorageHive`.